### ISSUE TITLE
Backslashes in Input Fields

### ISSUE THREAD

**Issue by muratpurc** [CONTENIDO/CONTENIDO#182]
Entered backslashes (\\) in backend form fields are being removed, the value is stored without the backslash.

Known areas are:
- System or client configuration
- Regular expressions for PIFA form field validation (e.g. "/\d{5}/")

See info in CONTENIDO forum:
https://forum.contenido.org/viewtopic.php?f=116&t=44475#p182151

**Comment by tone2tone** [comment]
Would it be enough to search for functions like  setSystemProperty() where client settings are stored and add a stripslashes() functionality to each value to store? 
For example, changing 
` $prop = $systemPropColl->setValueByTypeName($type, $name, $value);`
to
` $prop = stripslashes($systemPropColl->setValueByTypeName($type, $name, $value));`
or would you go into the setValueByTypeName function or even further down if necessary to add stripslashes to the lowest possible level?
We could also define a new function that just holds stripslashes for the time being, but ready to add further tweaks before storing values if necessary.
If I know which way to go I could dig a bit here.

**Comment by muratpurc** [comment]
@tone2tone  Slashes to all $_POST and $_GET variables may (it is not allways the case) be added in the globals_off.inc.php. This will be executed  in a very early stage of the CONTENIDO initialization. The reason for this is historical and a lot of logic in CONTENIDO relies on it. I would not change the behaviour on this level at the moment and have no idea, how to solve this issue.

**Comment by tone2tone** [comment]
Okay, got the problem wrong anyway after reading the forum post properly. Problem is not to REMOVE the slashes, but to keep them. Sorry for my confusion matters in the beginning. Need some time to adjust ;-) 
What I don't understand: In my 4.10.1, I try to save weird values with quotation marks and backslashes in the client settings area. globals_off.inc.php tells me that $cfg['simulate_magic_quotes'] is "true", yet, in names or values, no escaping happens. Backslashes are deleted and/or not added. Are the storing functions working differently and able to ignore the mq-settings?

**Comment by muratpurc** [comment]
@tone2tone 

What happens when you save system settings is:

1. globals_off.inc.php
All $_POST, $_GET, and $_COOKIE values will be processd by using addlashes (cString::addSlashes) if variable $cfg['simulate_magic_quotes'] and constant CON_STRIPSLASHES is set to true.

2. include.systemsettings.php
Values of the $_REQUEST variable will be processed by using stripslashes (cString::stripSlashes) if the CON_STRIPSLASHES is set to true.

I guess, this is the issue. The content of $_REQUEST won't be changed in globals_off.inc.php by using addlashes, but slashes will be removed in include.systemsettings.php.

Adding the following line in globals_off.inc.php should fix the issue:

`$_REQUEST = array_map(array('cString', 'addSlashes'), $_REQUEST);`

But this could lead to other unwanted side effects in other scripts, which work with $_REQUEST but don't prepare it like it is done in include.systemsettings.php. It could also lead to erros in custom modules and plugins. We can adapt the core code but this change may still lead to errors in custom modules and plugins.

Therefore, adding the line `$_REQUEST = array_map(...)` to globals_off.inc.php is not a good idea...

**Comment by tone2tone** [comment]
Perhaps we could revisit program code like the following example in include.clientsettings.php:

```
if (defined('CON_STRIPSLASHES')) {
    $request = cString::stripSlashes($_REQUEST);
} else {
    $request = $_REQUEST;
}
```
As you say, $_REQUEST is never changed, whatever state CON_STRIPSLASHES has. WHY on earth os $_REQUEST stripped here then, as if it wants to undo something that it has never done in the first place? IMHO it would suffice to cut the whole sequence down to
`$request = $_REQUEST;`
What do you think?

**Comment by muratpurc** [comment]
@tone2tone replacing the code in include.clientsettings.php with `$request = $_REQUEST;` should fix the problem. But this change should be checked, if the values in system settings are still stored and displayed correctly.

**Comment by tone2tone** [comment]
@purcaholic I added a pull request. In also tried to safe filter values in PIFA WITH backslashes and never came across problems, even without touching any code. I see no problem here and no need for intervention.
Side note: I also checked backslashes in TinyMCE4 windows. They are always additionally escaped, with or without proposed code changes. Given that backslashes here might make no sense at all, there might be no need to act on it. BUT if you think there is, let me know.
